# Exercício Faça você mesma

In [1]:
# import
import pandas as pd

## Leitura do CSV

In [2]:
df = pd.read_csv(
    "../data/licitacoes.csv",
    encoding="cp1252",
    sep=";"
)

## Licitações Encerradas

In [3]:
df[df['Situação Licitação']=="Encerrado"]

,Número Licitação,Código UG,Nome UG,Código Modalidade Compra,Modalidade Compra,Número Processo,Objeto,Situação Licitação,Código Órgão Superior,Nome Órgão Superior,Código Órgão,Nome Órgão,UF,Município,Data Resultado Compra,Data Abertura,Valor Licitação
3,12018,925135,BANCO CENTRAL DO BRASIL - RECIFE,6,Dispensa de Licitação,148097,"Objeto: TV LED de 65 polegadas, UHD 4k, com co...",Encerrado,20101,Sem informação,25201,BANCO CENTRAL DO BRASIL-ORC.FISCAL/SEG.SOCIAL,-3,Inválido,04/12/2018,30/11/2018,"11075,9900"
16,12018,130129,SUPERINT.REG.DE DESENV.DA LAV.CACAUEIRA/MT,6,Dispensa de Licitação,0000000000032018,Objeto: Atender despesas com telefonia fixa e ...,Encerrado,22000,Ministério da Agricultura e Pecuária,22000,Ministério da Agricultura e Pecuária - Unid,MT,ALTA FLORESTA,06/12/2018,NaN,"1541,7000"
18,12018,135097,EMBRAPA CAFE,7,Inexigibilidade de Licitação,21163.000694/2018,Objeto: Serviço de genotipagem de 384 amostras...,Encerrado,22000,Ministério da Agricultura e Pecuária,22202,Empresa Brasileira de Pesquisa Agropecuária,DF,BRASILIA,04/12/2018,NaN,"90504,0000"
19,12018,152140,IFB - CAMPUS TAGUATINGA NORTE,6,Dispensa de Licitação,23162004935201845,Objeto: Contratação de empesa especializada a ...,Encerrado,26000,Ministério da Educação,26428,Instituto Federal de Brasília,DF,BRASILIA,31/12/2018,NaN,"2068,0000"
24,12018,153092,BIBLIOTECA CENTRAL DA UFPE,6,Dispensa de Licitação,23076025958201862,Objeto: Aquisição de Equipamento de Proteção I...,Encerrado,26000,Ministério da Educação,26242,Universidade Federal de Pernambuco,PE,RECIFE,06/12/2018,29/11/2018,"5609,0000"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15327,907662018,926130,DIVISÃO DE CENTRO DE SERV COMPRATILHADOS BSB,6,Dispensa de Licitação,PL.CSG.A.90766,"Objeto: Fornecimento de fogão, para entrega na...",Encerrado,20101,Sem informação,91081,FURNAS-CENTRAIS ELETRICAS S.A.,-3,Inválido,10/12/2018,04/12/2018,"817,2000"
15328,907732018,926130,DIVISÃO DE CENTRO DE SERV COMPRATILHADOS BSB,6,Dispensa de Licitação,PL.CSG.A.90773,Objeto: Prestação de serviços para locação de ...,Encerrado,20101,Sem informação,91081,FURNAS-CENTRAIS ELETRICAS S.A.,-3,Inválido,20/12/2018,05/12/2018,"8400,0000"
15329,908252018,926130,DIVISÃO DE CENTRO DE SERV COMPRATILHADOS BSB,6,Dispensa de Licitação,PL.CSG.A.90825,Objeto: Coleta e descarte de resíduos industri...,Encerrado,20101,Sem informação,91081,FURNAS-CENTRAIS ELETRICAS S.A.,-3,Inválido,21/12/2018,17/12/2018,"9040,0000"
15333,990922018,158139,"INST.FED.DE EDUC.,CIENC.E TEC.FLUMINENSE",6,Dispensa de Licitação,23317009092201862,Objeto: Aquisição de material para revisão do ...,Encerrado,26000,Ministério da Educação,26434,Instituto Federal Fluminense,RJ,CAMPOS DOS GOYTACAZES,03/12/2018,NaN,"1149,7500"


In [4]:
df['Situação Licitação'].value_counts()

Situação Licitação
Encerrado                         9016
Publicado                         3709
Evento de Resultado de Julgame    1922
Retificação                        127
Evento de Alteração Publicad        93
Evento de Suspensão Publicado       72
Anulação                            67
Revogação                           63
Divulgado                           55
Evento de Adiamento Publicado       46
Evento de Retificação Public        45
Evento de Revogação Publicad        40
Evento de Alteração de Resul        37
Evento de Anulação Publicado        18
Evento de Retificação Divulg         7
Inválido                             4
Evento de Habilitação Public         4
Evento de Alteração Divulgad         4
Evento de Reabertura de Prazo        4
Evento de Adiamento Divulgado        3
Name: count, dtype: int64

## Verificando e Tratando Valores Faltantes

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15336 entries, 0 to 15335
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Número Licitação          15336 non-null  int64
 1   Código UG                 15336 non-null  int64
 2   Nome UG                   15336 non-null  str  
 3   Código Modalidade Compra  15336 non-null  int64
 4   Modalidade Compra         15336 non-null  str  
 5   Número Processo           15336 non-null  str  
 6   Objeto                    15336 non-null  str  
 7   Situação Licitação        15336 non-null  str  
 8   Código Órgão Superior     15336 non-null  int64
 9   Nome Órgão Superior       15336 non-null  str  
 10  Código Órgão              15336 non-null  int64
 11  Nome Órgão                15336 non-null  str  
 12  UF                        15336 non-null  str  
 13  Município                 15336 non-null  str  
 14  Data Resultado Compra     15336 non-null  str  
 

A tabela tem 15336 linhas de dados, na coluna "Data Abertura" há apeans 4688 linhas não nulas, cerca de 30,5% dos dados. 

O tipo de dado da coluna é string (str), pois as datas estão armazenadas em formato textual.

Como o enunciado permite substituir os valores ausentes por datas inexistentes, os valores faltantes foram preenchidos com "00/00/0000", mantendo o mesmo formato dos demais registros da coluna.

In [8]:
df.loc[df['Data Abertura'].isnull(), 'Data Abertura'] = "00/00/0000"

In [9]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15336 entries, 0 to 15335
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype
---  ------                    --------------  -----
 0   Número Licitação          15336 non-null  int64
 1   Código UG                 15336 non-null  int64
 2   Nome UG                   15336 non-null  str  
 3   Código Modalidade Compra  15336 non-null  int64
 4   Modalidade Compra         15336 non-null  str  
 5   Número Processo           15336 non-null  str  
 6   Objeto                    15336 non-null  str  
 7   Situação Licitação        15336 non-null  str  
 8   Código Órgão Superior     15336 non-null  int64
 9   Nome Órgão Superior       15336 non-null  str  
 10  Código Órgão              15336 non-null  int64
 11  Nome Órgão                15336 non-null  str  
 12  UF                        15336 non-null  str  
 13  Município                 15336 non-null  str  
 14  Data Resultado Compra     15336 non-null  str  
 

In [10]:
df

,Número Licitação,Código UG,Nome UG,Código Modalidade Compra,Modalidade Compra,Número Processo,Objeto,Situação Licitação,Código Órgão Superior,Nome Órgão Superior,Código Órgão,Nome Órgão,UF,Município,Data Resultado Compra,Data Abertura,Valor Licitação
0,12016,160438,COMANDO 2 BRIGADA DE CAVALARIA MECANIZADA,-99,Pregão - Registro de Preço,80871182401201629,Objeto: Pregão Eletrônico - Aquisição de mater...,Evento de Resultado de Julgame,52000,Ministério da Defesa,52121,Comando do Exército,RS,URUGUAIANA,04/12/2018,17/08/2016,"18854,0000"
1,12017,160203,2º BATALHAO DE ENGENHARIA DE CONSTRUCAO,3,Concorrência,64040053530201713,Objeto: Contratação de empresa para construção...,Evento de Resultado de Julgame,52000,Ministério da Defesa,52121,Comando do Exército,PI,TERESINA,07/12/2018,00/00/0000,"889571,4300"
2,12017,193002,DEPARTAMENTO NACIONAL DE OBRAS CONT.AS SECAS,2,Tomada de Preços,59400003079201636,Objeto: Contratação de empresa especializada n...,Evento de Resultado de Julgame,53000,Ministério da Integração e do Desenvolvime,22204,Departamento Nacional de Obras Contra as Seca,CE,FORTALEZA,07/12/2018,00/00/0000,"865536,6500"
3,12018,925135,BANCO CENTRAL DO BRASIL - RECIFE,6,Dispensa de Licitação,148097,"Objeto: TV LED de 65 polegadas, UHD 4k, com co...",Encerrado,20101,Sem informação,25201,BANCO CENTRAL DO BRASIL-ORC.FISCAL/SEG.SOCIAL,-3,Inválido,04/12/2018,30/11/2018,"11075,9900"
4,12018,110101,GABINETE DA VICE-PRESIDENCIA DA REPUBLICA,7,Inexigibilidade de Licitação,00300000433201818,Objeto: Pretação de serviços de publicação de ...,Publicado,20000,Presidência da República,60000,Gabinete da Vice-Presidência da República,DF,BRASILIA,21/12/2018,00/00/0000,"12000,0000"
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
15331,920182018,135023,EMBRAPA/CNPAB,-99,Pregão - Registro de Preço,21150000453201820,Objeto: Pregão Eletrônico - Contratar Empresa ...,Evento de Resultado de Julgame,22000,Ministério da Agricultura e Pecuária,22202,Empresa Brasileira de Pesquisa Agropecuária,RJ,SEROPEDICA,04/12/2018,26/11/2018,"2588060,0800"
15332,924192017,158274,INST.FED. FLUMINENSE/CAMPUS CAMPOS-CENTRO,6,Dispensa de Licitação,23318002419201784,Objeto: Retificação da dispensa de licitação n...,Publicado,26000,Ministério da Educação,26434,Instituto Federal Fluminense,RJ,CAMPOS DOS GOYTACAZES,26/12/2018,00/00/0000,"490325,0000"
15333,990922018,158139,"INST.FED.DE EDUC.,CIENC.E TEC.FLUMINENSE",6,Dispensa de Licitação,23317009092201862,Objeto: Aquisição de material para revisão do ...,Encerrado,26000,Ministério da Educação,26434,Instituto Federal Fluminense,RJ,CAMPOS DOS GOYTACAZES,03/12/2018,00/00/0000,"1149,7500"
15334,996252018,153035,UNIVERSIDADE FEDERAL DO TRIANGULO MINEIRO,6,Dispensa de Licitação,23085000625201811,Objeto: Pagamento da taxa referente ao primeir...,Encerrado,26000,Ministério da Educação,26254,Universidade Federal do Triângulo Mineiro,MG,UBERABA,04/12/2018,00/00/0000,"298,0000"


## o